ompi/runtime/ompi_spc.c

```diff
@@ -0,0 +1,510 @@
+        index = pvar->pvar_index - mpi_t_offset;
+        attached_event[index] = 1;
+    }
+    /* For this event, we need to turn off the counter */
+    else if(MCA_BASE_PVAR_HANDLE_STOP == event) {
+        /* Convert from MPI_T pvar index to SPC index */
+        index = pvar->pvar_index - mpi_t_offset;
+        attached_event[index] = 0;
+    }
+
+    return MPI_SUCCESS;
+}
+
+/* ##############################################################
+ * ################# Begin SPC Functions ########################
+ * ##############################################################
+ */
+
+/* This function returns the current count of an SPC counter that has been retistered
+ * as an MPI_T pvar.  The MPI_T index is not necessarily the same as the SPC index,
+ * so we need to convert from MPI_T index to SPC index and then set the 'value' argument
+ * to the correct value for this pvar.
+ */
+static int ompi_spc_get_count(const struct mca_base_pvar_t *pvar, void *value, void *obj_handle)
+{   
+    (void) obj_handle;
+
+    long long *counter_value = (long long*)value;
+
+    if(OPAL_UNLIKELY(mpi_t_disabled == 1)) {
+        *counter_value = 0;
+        return MPI_SUCCESS;
+    }
+
+    /* Convert from MPI_T pvar index to SPC index */
+    int index = pvar->pvar_index - mpi_t_offset;
+    /* Set the counter value to the current SPC value */
+    *counter_value = events[index].value;
+    /* If this is a timer-based counter, convert from cycles to microseconds */
+    if(timer_event[index])
+        *counter_value /= sys_clock_freq_mhz;
+    /* If this is a high watermark counter, reset it after it has been read */
+    if(index == OMPI_MAX_UNEXPECTED_IN_QUEUE || index == OMPI_MAX_OOS_IN_QUEUE)
+        events[index].value = 0;
+
+    return MPI_SUCCESS;
+}
+
+/* Initializes the events data structure and allocates memory for it if needed. */
+void events_init()
+{
+    int i;
+
+    /* If the events data structure hasn't been allocated yet, allocate memory for it */
+    if(events == NULL) {
+        events = (ompi_event_t*)malloc(OMPI_NUM_COUNTERS * sizeof(ompi_event_t));
+    }
+    /* The data structure has been allocated, so we simply initialize all of the counters
+     * with their names and an initial count of 0.
+     */
+    for(i = 0; i < OMPI_NUM_COUNTERS; i++) {
+        events[i].name = counter_names[i];
+        events[i].value = 0;
+    }
+}
+
+/* Initializes the SPC data structures and registers all counters as MPI_T pvars.
+ * Turns on only the counters that were specified in the mpi_spc_attach MCA parameter.  
+ */
+void ompi_spc_init()
+{
+    int i, j, ret, found = 0, all_on = 0;
+
+    /* Initialize the clock frequency variable as the CPU's frequency in MHz */
+    sys_clock_freq_mhz = opal_timer_base_get_freq() / 1000000;
+
+    events_init();
+
+    /* Get the MCA params string of counters to turn on */
+    char **arg_strings = opal_argv_split(ompi_mpi_spc_attach_string, ',');
```

Comment:
Don't forget to `opal_argv_free(arg_strings)` when you leave this function.

Reply:
It has been fixed.